### ISSUE TITLE
Arrows from a box back to itself don't look so good

### ISSUE THREAD

**Issue by airhorns** [steveruizok/perfect-arrows#9]
Definitely for state machines and potentially for other use cases as well it'd be awesome to have support for arrows that go from a box back to itself. This would represent a transition from one state to the same state in a state machine-y world which is super valid if you ask me. Here's what happens now:

<img width="519" alt="Screen Shot 2020-09-18 at 8 41 09 AM" src="https://user-images.githubusercontent.com/158950/93598657-ff2a8080-f98a-11ea-8339-71e0353efc74.png">

I would be happy to work on (or get someone to work on) this but I think it'd be cool to agree on what the perfect version of an arrow of this nature would look like! My thoughts:

 - in the event that there are multiple self arrows, it'd be good to hint to the algorithm to put them in different places so they don't overlap. This is possible for box to box arrows with `flip` and `bow` right now
 - it'd be good to be able to adjust the style (padding etc) to match whatever you might be doing with the other box to box arrows
 - probably super out of scope for now, but it'd be really great to somehow ask for as few crossing arrows as possible for aesthetics. In the meantime, maybe we'd want self-arrows to bias towards start points on the boxes that the box-to-box arrows bias away from. 

My suggestion would be a perfect circle anchored at one of the corners of the box, with the same start and end padding behaviour as the box-to-box function, and maybe with its centre offset out a little along the ray intersecting the anchor corner and the centre of the box. 

I'm also not sure if the preferred way to do this is a new public function, or adding a check to the box-to-box function to see if the boxes are the same. I'd suggest a new function to disambiguate between transitions from one box to itself, and from one to another box that is perfectly on top of it.

I added a test in the `example` here if you'd like to play with the current implementation's output: https://github.com/gadget-inc/perfect-arrows/tree/box-to-self

**Comment by steveruizok** [comment]
Hey, I know what you mean! I started writing perfect-arrows for my [state-charts app](https://ide.state-designer.com) and this was one of the cases I started from.

Here's the issue. Given the limited inputs to `getBoxToBoxArrow`, the function would have no way of knowing whether the arrow is pointing to the _same_ box or to a _different_ box that happens to have an identical position and size. Really, in order to know that, we'd have to know some kind of context of boxes. If we knew which boxes existed and which arrows connected which boxes, then we could identify self-to-self boxes and work on avoiding crossings, too. But all of that is a level up from what this library is aiming for, though it could build on some of the solutions that we are working on here. Maybe `perfect-directed-graph`? :)

So instead, I think this would need to be its own function, `getBoxToSelfArrow`. My back-of-the-napkin API would be something like:

```ts
function getBoxToSelfArrow(
  x: number, 
  y: number, 
  w: number, 
  h: number, 
  options: {
    radius?: number // How big to make the loop
    angle?: number, // Where on the box the loop should start and end
    padStart?: number
    padEnd?: number
    flip?: boolean
  }
) {
  ...
  return [sx, sy, cx, cy, ex, ey, ae, as, ac]
}
```

![image](https://user-images.githubusercontent.com/23072548/93616243-661c5980-f9cc-11ea-94da-2197d0af498a.png)

The return values could be the same. You could then draw an ellipse that starts and ends at the given angles.

```ts
const [sx, sy, cx, cy, ex, ey, ae, as] = getBoxToSelfArrow(0,0,100,100)

ctx.moveTo(sx, sy)
ctx.arc(cx, cy, radius, as, ae)
```

If you did have that higher level "box manager" in your app, I think it could adjust these settings to accommodate a box with multiple self-to-self loops by setting a different angle.


```ts
for (let i = 0; i < loops.length; i++) {
  const [sx, sy, cx, cy, ex, ey, ae, as] = getBoxToSelfArrow(0,0,100,100, {
    angle: (Math.PI / 8) * i
  })
```

Or you could just decide to use this function instead if you had a perfect overlap.

```ts
if (boxesPerfectlyOverlap(x0, y0, w0, h0, x1, y1, w1, h1) {
  getBoxToSelfArrow(x0, y0, w0, h0)
} else {
  getBoxToBoxArrow(x0, y0, w0, h0, x1, y1, w1, h1)
}
```

**Comment by steveruizok** [comment]
Still working on this one. The newest update makes identical box-to-box arrows potentially worse, so be sure to check in your code for whether you need an arrow. Until we have a solution, I'd recommend not drawing an arrow between boxes if they are perfectly centered.

**Comment by airhorns** [comment]
@steveruizok is this something that you're planning on tackling sometime soon? No worries if not -- I can start work on a PR! If you have something half baked or almost done I would love to help push it past the finish line!

**Comment by steveruizok** [comment]
Go for it! I did some initial experimentation with box-to-self arrows and the problem is (as expected) more tricky than I originally thought. I've decided to work on the polygon-to-polygon problem next, so if you'd like to take this one on I'd be excited to see what you come up with.

My only thought would be that the output should be the same as with other functions (returning start, center, end, etc.) though in this case the center should be the center of the circle segment rather than the corner of a quadratic curve.

Good luck!
